metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java

```diff
@@ -0,0 +1,262 @@
+            this.minBytesSinceLastSnapshot = minBytesSinceLastSnapshot;
+            return this;
+        }
+
+        public Builder setMinTimeSinceLastSnapshotNs(long minTimeSinceLastSnapshotNs) {
+            this.minTimeSinceLastSnapshotNs = minTimeSinceLastSnapshotNs;
+            return this;
+        }
+
+        public SnapshotGenerator build() {
+            if (emitter == null) throw new RuntimeException("You must set a snapshot emitter.");
+            return new SnapshotGenerator(
+                nodeId,
+                time,
+                emitter,
+                faultHandler,
+                minBytesSinceLastSnapshot,
+                minTimeSinceLastSnapshotNs
+            );
+        }
+    }
+
+    /**
+     * The callback which actually generates the snapshot.
+     */
+    public interface Emitter {
+        void emit(MetadataImage image);
+    }
+
+    /**
+     * The node ID.
+     */
+    private final int nodeId;
+
+    /**
+     * The clock to use.
+     */
+    private final Time time;
+
+    /**
+     * The emitter callback, which actually generates the snapshot.
+     */
+    private final Emitter emitter;
+
+    /**
+     * The slf4j logger to use.
+     */
+    private final Logger log;
+
+    /**
+     * The fault handler to use.
+     */
+    private final FaultHandler faultHandler;
+
+    /**
+     * The minimum number of bytes we will wait to see before emitting a snapshot.
+     */
+    private final long minBytesSinceLastSnapshot;
+
+    /**
+     * The minimum amount of time we will wait before emitting a snapshot, or 0 to disable
+     * time-based snapshotting.
+     */
+    private final long minTimeSinceLastSnapshotNs;
+
+    /**
+     * If non-null, the reason why snapshots have been disabled.
+     */
+    private volatile String disabledReason;
+
+    /**
+     * The event queue used to schedule emitting snapshots.
+     */
+    private final EventQueue eventQueue;
+
+    /**
+     * The log bytes that we have read since the last snapshot.
+     */
+    private long bytesSinceLastSnapshot;
+
+    /**
+     * The time at which we created the last snapshot.
+     */
+    private long lastSnapshotTimeNs;
+
+    private SnapshotGenerator(
+        int nodeId,
+        Time time,
+        Emitter emitter,
+        FaultHandler faultHandler,
+        long minBytesSinceLastSnapshot,
+        long minTimeSinceLastSnapshotNs
+    ) {
+        this.nodeId = nodeId;
+        this.time = time;
+        this.emitter = emitter;
+        this.faultHandler = faultHandler;
+        this.minBytesSinceLastSnapshot = minBytesSinceLastSnapshot;
+        this.minTimeSinceLastSnapshotNs = minTimeSinceLastSnapshotNs;
+        LogContext logContext = new LogContext("[SnapshotGenerator " + nodeId + "] ");
+        this.log = logContext.logger(SnapshotGenerator.class);
+        this.disabledReason = null;
+        this.eventQueue = new KafkaEventQueue(time, logContext, "");
```

Comment:
Perhaps we should add a thread name prefix?

Reply:
fair enough. we have all the information here, so there is no reason why not (and some good reasons why to!)